docs/source/fine-performance-metrics.rst

```diff
@@ -0,0 +1,198 @@
+   :alt: Fine Performance Metrics dashboard, except the I/O functions
+
+This tells us an important piece of information: why is `astype`, which is a pure CPU
+function, spending so much time occupying workers' threads but without accruing any CPU
+time? The answer, almost certainly, it's that it doesn't properly release the GIL.
+
+Fine Performance Metrics don't collect exclusively wall clock timings. We can change the
+unit to bytes:
+
+.. image:: images/fine-performance-metrics/bytes.png
+   :alt: Fine Performance Metrics dashboard, unit=bytes
+
+The above gives us insights on spill/unspill activity (see :doc:`worker-memory`): in
+this workflow, in 99% of the cases there was enough RAM to accommodate all the data and
+it was unnecessary to retrieve it from disk; in other words we have a 99% *cache hit
+ratio*, which means that we would not get any benefit if we increased RAM, but we would
+likely start seeing slowdowns if we reduced it.
+
+
+Are task prefixes enough?
+-------------------------
+Individual task prefixes may be too granular; vice versa, the same task prefix may
+appear in very different sections of the workflow. Your codebase may be complex enough
+that it's not straightforward to pinpoint the client code that's responsible for a
+specific task prefix.
+
+:doc:`spans` let you break down these metrics into macro-blocks (e.g. data load,
+preprocessing, etc.).
+
+
+API for advanced users
+----------------------
+For the most part, fine performance metrics just work; as a user, you don't need to
+change your client code.
+
+If you run custom tasks on the cluster (e.g. via :meth:`~distributed.Client.submit`,
+:meth:`~dask.array.Array.map_blocks`, or
+:meth:`~dask.dataframe.DataFrame.map_partitions`), you may want to customize the metrics
+they produce. For example, you may want to separate I/O time from `thread-noncpu`:
+
+.. code-block:: python
+
+   from distributed.metrics import context_meter
+
+   @context_meter.meter("I/O")
+   def read_some_files():
+       ...
+
+   future = client.submit(read_some_files)
+
+In the above example, the wall time spent by the custom function ``read_some_files``
+will be logged as "I/O", which is a completely arbitrary activity label.
+
+Alternatively you may want to just label some of the time this way:
+
+.. code-block:: python
+
+   def read_some_files():
+       with context_meter.meter("I/O"):
+           data = read_from_network(...)
+       return preprocess(data)
+
+In the above example, the function is split into an I/O intensive phase,
+``read_from_network``, and a CPU-intensive one, ``preprocess``. The
+``context_meter.meter`` context manager will subtract the time spent by
+``read_from_network`` from ``thread-noncpu``, whereas the time spent by ``preprocess``
+will still be a mix of ``thread-cpu`` and ``thread-noncpu`` (the latter may, for
+example, highlight GIL contention).
+
+.. note::
+   The ``context_meter.meter`` context manager wraps around code that runs on the
```

Comment:
Can we link to these methods? E.g. are there dosctrings available?

Reply:
done